client/health_test.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client_test
+
+import (
+	"net/url"
+
+	"gopkg.in/check.v1"
+
+	"github.com/canonical/pebble/client"
+)
+
+func (cs *clientSuite) TestHealthGet(c *check.C) {
+	cs.rsp = `{
+		"type": "sync",
+		"status-code": 200,
+		"status": "OK",
+		"result": {
+			"healthy": true
+		}
+	}`
+
+	opts := client.HealthOptions{
+		Level: client.AliveLevel,
+		Names: []string{"chk1", "chk3"},
+	}
+	health, err := cs.cli.Health(&opts)
+	c.Assert(err, check.IsNil)
+	c.Assert(health, check.DeepEquals, &client.HealthInfo{
+		Healthy: true,
+	})
+	c.Assert(cs.req.Method, check.Equals, "GET")
+	c.Assert(cs.req.URL.Path, check.Equals, "/v1/health")
+	c.Assert(cs.req.URL.Query(), check.DeepEquals, url.Values{
+		"level": {"alive"},
+		"names": {"chk1", "chk3"},
+	})
+}
+
+func (cs *clientSuite) TestHealthDefaultOptions(c *check.C) {
+	cs.rsp = `{
+		"type": "sync",
+		"status-code": 502,
+		"status": "Bad Gateway",
```

Comment:
The `/v1/health` endpoint returns a "Bad Gateway" status with 502 code when unhealthy.

```json
{"type":"sync","status-code":502,"status":"Bad Gateway","result":{"healthy":false}}
```

Reply:
Yeah, we (I can't remember who exactly) debated which HTTP status code this should return at the time, and ended up with 502. It's nice a perfect fit, but we wanted it to be a 5xx code, and Pebble is kind of acting as a proxy/gateway here, so it seemed like a reasonable one.

Reply:
This seems a bit awkward. Generally those errors mean "I can't handle your request", because of reason 5XX. The request here is "How's your health?", so in theory the response should be "Yep, the check you're asking for is not ready/alive.".

This seems analogous to asking to pull a file and getting a 404 to say the file doesn't exist. That's a bad practice as we're mixing the success of the request with the result of the operation requested. We would get a similar 404 if we were talking to a different server altogether, for example.

Maybe this was done to satisfy Kubernetes, when poking at health URLs using its internal probing?